### DISCUSSION TITLE
Add `Either` to the standard library?

### DISCUSSION THREAD

**Discussion by stephentetley** [flix/flix#5873] [Ideas]
Should we add `Either` to the stdlib?

The case for isn't overwhelming as we already have `Result` which covers a lot of the common uses, but I've been looking at Arrows, Selective Functors and Profunctors which all use `Either` in their families of classes. It would nice to have a standard, canonical implementation of `Either`.

**Comment by magnus-madsen** [comment]
I am on the fence.

Points for:
- One sometimes wants to return either "this" or "that" where using `Result` feels wrong because neither result is "Ok" or "Err".

Points against:
- It is isomorphic to Result.
- It could cause confusion. When to use which (?) Possibly a name like `enum Xor[a, b] { case Fst(a), Snd(b) }` would help with that.

Questions:
- Would it implement type classes in a left-biased way? E.g. would it implement `Functor`? 

If it does not there is less potential for confusion, but it is also much less useful.

It seems Rust has `Result`, but not `Either`. 

Finally, it would not be too hard to put into its own package. All the infrastructure is there.

**Comment by magnus-madsen** [comment]
https://docs.rs/either/latest/either/enum.Either.html

I am leaning towards it being its own library that you can depend on. 

EDIT: Although it has 100M downloads (!?) https://crates.io/crates/either

**Comment by stephentetley** [comment]
Thanks Magnus, when I've finished `instanceof` I'll look to making `Either` as it's own library. It shouldn't need dependencies so could go in the community build for visibility.

It can always go in the stdlib later if there's a consensus to include it.

**Comment by jaschdoc** [comment]
If we want batteries to be included for Flix and to allow for different styles of programming, e.g., functional but with imperative implementation, it seems like a natural choice to have an `Either` type known from other functional languages, so people used to functional languages can immediately start using the types they're used to.
